File: src/content/docs/topics/cs/coding-problems/linked-list/024-swap-nodes-in-pairs-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val: number = 0, next: ListNode | null = null) {
        this.val = val; this.next = next;
    }
}

function toList(head: ListNode | null): number[] {
    const out: number[] = [];
    while (head) { out.push(head.val); head = head.next; }
    return out;
}

function fromList(vals: number[]): ListNode | null {
    const dummy = new ListNode();
    let cur = dummy;
    for (const v of vals) { cur.next = new ListNode(v); cur = cur.next; }
    return dummy.next;
}

function swapPairs(head: ListNode | null): ListNode | null {
    const dummy = new ListNode(0, head);
    let prev: ListNode = dummy;
    let cur: ListNode | null = head;
    while (cur && cur.next) {
        const nextPair = cur.next.next;
        prev.next = cur.next;
        cur.next.next = cur;
        cur.next = nextPair;
        prev = cur;
        cur = nextPair;
    }
    return dummy.next;
}

assert(JSON.stringify(toList(swapPairs(fromList([1,2,3,4])))) === JSON.stringify([2,1,4,3]));
assert(JSON.stringify(toList(swapPairs(fromList([])))) === JSON.stringify([]));
assert(JSON.stringify(toList(swapPairs(fromList([1])))) === JSON.stringify([1]));
assert(JSON.stringify(toList(swapPairs(fromList([1,2,3])))) === JSON.stringify([2,1,3]));
console.log('all tests pass');

File: src/content/docs/topics/cs/coding-problems/linked-list/024-swap-nodes-in-pairs-approach2.py
from __future__ import annotations

class ListNode:
    def __init__(self, val: int = 0, next: ListNode | None = None) -> None:
        self.val = val
        self.next = next

def to_list(head: ListNode | None) -> list[int]:
    out = []
    while head:
        out.append(head.val)
        head = head.next
    return out

def from_list(vals: list[int]) -> ListNode | None:
    dummy = ListNode()
    cur = dummy
    for v in vals:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next

def swap_pairs(head: ListNode | None) -> ListNode | None:
    dummy = ListNode(0, head)
    prev, cur = dummy, head
    while cur and cur.next:
        next_pair = cur.next.next
        prev.next = cur.next
        cur.next.next = cur
        cur.next = next_pair
        prev = cur
        cur = next_pair
    return dummy.next

assert to_list(swap_pairs(from_list([1,2,3,4]))) == [2,1,4,3]
assert to_list(swap_pairs(from_list([]))) == []
assert to_list(swap_pairs(from_list([1]))) == [1]
assert to_list(swap_pairs(from_list([1,2,3]))) == [2,1,3]
print("all tests pass")
